### PR TITLE
feat: migrate logic from ClickUp to Jira #PLTM-233

### PR DESCRIPTION
**Pull request by kwaszczuk**
## Changes

- enhance parsing of task references
  - use simpler, iterative implementation
  - remove parsing ClickUp's custom task ids (`#20jt35r`-like)
- remove for ClickUp tickets
- add logic for Jira issues
  - fetch issue existence and its name from Jira REST API

## Testing

Tested on two PRs https://github.com/RampNetwork/ramp-infra/pull/779 and https://github.com/RampNetwork/ramp-instant/pull/20764. For both Jira references were detected correctly.

### CODE REVIEW

File: dangerfile.js [hunk 1]
```diff
@@ -2,103 +2,72 @@ const axios = require('axios');
 
 const getTasks = require('./lib/get-tasks.js');
 
-const CLICKUP_API_URL = 'https://api.clickup.com/api/v2';
-const CLICKUP_TEAM_ID = '24301226';
+const JIRA_BASE_URL = 'https://rampnetwork.atlassian.net';
+const JIRA_API_URL = `${JIRA_BASE_URL}/rest/api/2`;
+const JIRA_TOKEN_USER = 'michal.jasiorowski@ramp.network'
```
Comment (kwaszczuk) [inline dangerfile.js]: Jira has personal tokens only, thus it is bounded to the specific user and their email needs to be provided during authentication. The approach is agreed with @MichalJasiorowski.

Comment (quezak) [reply dangerfile.js]: we should create a separate tech account for this purpose, like we had the CodeMom account on clickup.

Comment (quezak) [reply dangerfile.js]: and the username should be passed in args along with the token, not hardcoded

File: dangerfile.js [hunk 2]
```diff
@@ -2,103 +2,72 @@ const axios = require('axios');
...
+    return [];
   }
 
   return Promise.all(tasks.map(req));
 };
 
-const checkAndUpdateClickupIssues = async () => {
-  const source = [danger.github.pr.title, danger.github.pr.body].join(' ');
+const checkTasks = async () => {
+  const source = [danger.github.pr.title].join(' ');
```
Comment (quezak) [inline dangerfile.js]:
```suggestion
  const source = danger.github.pr.title;
```

File: lib/get-tasks.js [hunk 3]
```diff
@@ -1,27 +1,25 @@
...
-    const ticketRefRegex = /(?<=\s|^)#(([A-Z]{2,10}-\d+)|(\w{7,10}))/g;
-    const ticketIdRegex = /#(?<taskId>([A-Z]{2,10}-\d+)|(\w{7,10}))/;
-    const customTicketIdRegex = /#[A-Z]{2,10}-\d+/;
-    const ids =
-      source.match(ticketRefRegex)?.map(v => ({
-        ...v.match(ticketIdRegex).groups,
-        isCustom: customTicketIdRegex.test(v),
-      })) || [];
+  const taskRefRegex = /(?<=\s|^)#[A-Z]{2,10}-\d+/g;
+  const refs = source.match(taskRefRegex) || [];
```
Comment (quezak) [inline lib/get-tasks.js]: We agreed to put the task ids at the end, so let's also check that to not have chaos :D
To do that and still be able to catch multiple tasks, we need to match the whole task ids string as one and split it:
image
```suggestion
  const taskRefRegex = /(\s+#[A-Z]{2,10}-\d+)+\s*$/;
  const allTaskIds = taskRefRegex.match(source)?.[0]?.trim() ?? '';
  const refs = allTaskIds.split(/\s+/);
```

File: lib/get-tasks.js [hunk 4]
```diff
@@ -1,27 +1,25 @@
...
+    tasks.push({
+      taskId: ref.slice(1), // remove # from the beginning of the reference
+    });
+  }
+  // sort tasks alphabetically
+  tasks.sort(function(refA, refB) {
+    if (refA.taskId < refB.taskId) return -1;
+    if (refA.taskId > refB.taskId) return 1;
+    return 0;
+  });
```
Comment (quezak) [inline lib/get-tasks.js]: is this needed? IMO we should keep tasks in order they appeared in the title

File: lib/get-tasks.spec.js [hunk 5]
```diff
@@ -2,51 +2,44 @@ const getTasks = require('./get-tasks.js');
...
-    "This PR closes #86bwmzcyf",
+    "This PR has duplicated reference #BUNNY-1 #BUNNY-1",
     [
-      { taskId: "86bwmzcyf", isCustom: false }
+      { taskId: "BUNNY-1" },
     ]
   ],
   [
-    "This PR closes #86bwmzcyf but it's in the middle of the text",
+    "This PR closes #CORE-123 and is related to #BUNNY-1",
```
Comment (quezak) [inline lib/get-tasks.spec.js]: are these tests relevant when we only check PR title, not description? The title has a structured format (see e.g. my regex101 screenshot in prev comment)

File: lib/get-tasks.js (hunk 4 above)
Comment (kwaszczuk) [reply lib/get-tasks.js]: Probably not, just my OCD wanted everything to be sorted 😅 I've now removed it to have a simpler logic 4fc47c6a790f3407443c43b56859ac028bb99d53.

File: lib/get-tasks.spec.js (hunk 5 above)
Comment (kwaszczuk) [reply lib/get-tasks.spec.js]: Matching function tested here does not have any assumptions regarding the input format, thus I don't think we really need to adjust test cases to match the real use case. We can just focus on testing its functionality, that is detecting issue reference at the end of the input.

File: lib/get-tasks.js (hunk 3 above)
Comment (kwaszczuk) [reply lib/get-tasks.js]: At first I've liked the idea and implemented it 94a88d1c2bacb8cfbf5c82c0984e89a414a40dc6. But now I've realized there are some repos where regularly task reference was added at the beginning of the title e.g. [RampNetwork/data-eng-core](https://github.com/RampNetwork/data-eng-core/pulls) or [RampNetwork/ramp-infra](https://github.com/RampNetwork/ramp-infra/pulls).

I am afraid enforcing issue reference at the end of title for all repositories may bring some chaos, what do you think?

Comment (quezak) [reply lib/get-tasks.js]: I don't have access to data-eng-core, but looking at the infra examples I'd say putting the task id in the front is incorrect, as the PR title should also stick to the conventional commits convention with the change type & scope at the beginning. So I actually think that it's good to enforce this.

Comment (quezak) [reply lib/get-tasks.js]: we may extend the error message to include that the task ids must be "at the end" of PR title.

File: lib/get-tasks.spec.js (hunk 5 above)
Comment (quezak) [reply lib/get-tasks.spec.js]: yeah but if we change the regex as I'm proposing then it will be invalid :D

Comment (kwaszczuk) [reply lib/get-tasks.spec.js]: True, I've already adjusted tests 10a4277c943f3cd19d7742b5eabc76420e70ae35

File: lib/get-tasks.js (hunk 3 above)
Comment (kwaszczuk) [reply lib/get-tasks.js]: Done c6c277eeffdbb6b29f9b87d3d6d332f5ecee5062

File: dangerfile.js (hunk 1 above)
Comment (kwaszczuk) [reply dangerfile.js:7]: I've discussed an idea of using Codemon account for Clickup and that's fine. The respective Jira account should be provisioned in the upcoming days.

Comment (kwaszczuk) [reply dangerfile.js]: Token's user is know provided via Github variable: 332349a27c14ba0a0a27881699dd3ad28bc4e12a

**Review by quezak — APPROVED** [review]

**Pull request merged**